io-engine/src/grpc/v1/replica.rs

```diff
@@ -492,4 +573,99 @@ impl ReplicaRpc for ReplicaService {
+                            Some(bdev) => Lvol::try_from(bdev)?,
+                            None => {
+                                return Err(LvsError::Invalid {
+                                    source: Errno::ENOENT,
+                                    msg: format!(
+                                        "Replica {} not found",
+                                        args.replica_uuid
+                                    ),
+                                })
+                            }
+                        };
+                        let snapshots = lvol
+                            .list_snapshot()
+                            .into_iter()
+                            .map(ReplicaSnapshot::from)
+                            .collect();
+                        Ok(ListReplicaSnapshotsResponse {
+                            snapshots,
+                        })
+                    } else {
+                        let snapshots = Lvol::list_all_snapshots()
+                            .into_iter()
+                            .map(ReplicaSnapshot::from)
+                            .collect();
+                        Ok(ListReplicaSnapshotsResponse {
+                            snapshots,
+                        })
+                    }
+                })?;
+                rx.await
+                    .map_err(|_| Status::cancelled("cancelled"))?
+                    .map_err(Status::from)
+                    .map(Response::new)
+            },
+        )
+        .await
+    }
+
+    #[named]
+    async fn delete_replica_snapshot(
+        &self,
+        request: Request<DeleteReplicaSnapshotRequest>,
+    ) -> GrpcResult<()> {
+        self.locked(
+            GrpcClientContext::new(&request, function_name!()),
+            async move {
+                let args = request.into_inner();
+                info!("{:?}", args);
+                let rx = rpc_submit(async move {
+                    let bdev = UntypedBdev::bdev_first().expect("Failed to enumerate devices");
+
+                    let mut devices = bdev
+                        .into_iter()
+                        .filter(|b| b.driver() == "lvol" && b.uuid_as_string() == args.snapshot_uuid)
+                        .map(|b| Lvol::try_from(b).unwrap())
+                        .collect::<Vec<Lvol>>();
+                    // Fail operation if snapshot not found or multiple instance
+                    // found
+                    if devices.len() > 1 || devices.is_empty() {
+                        return Err(LvsError::Invalid {
+                                source: Errno::ENOENT,
+                                msg: format!(
+                                    "Snapshot {} not found or multiple instance of snapshot with same name present",
+                                    args.snapshot_uuid
+                                ),
+                        })
+                    }
+                    devices.remove(0).destroy().await?;
```

Comment:
On restart, do snapshots get loaded as soon as parent replica is loaded?

Reply:
> On restart, do snapshots get loaded as soon as parent replica is loaded?

The bahaviour for snapshot must be same as basic replica. As post restart, all replica loaded back, same should happen for snapshot as well as both are lvol from SPDK point of view. Need a test verification to double sure on this.

Reply:
Thanks for confirming :+1: